pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -224,6 +151,269 @@ func (a *AmbientIndex) matchesScope(scope model.WaypointScope, w *model.Workload
 	return true
 }
 
+func PickBestPod(pods []*v1.Pod) *v1.Pod {
+	var best *v1.Pod
+	for _, p := range pods {
+		if !IsPodRunning(p) {
+			continue
+		}
+		if best == nil || p.CreationTimestamp.After(best.CreationTimestamp.Time) {
+			best = p
+		}
+	}
+	return best
+}
+
+func (a *AmbientIndex) Recompute(name types.NamespacedName) *model.WorkloadInfo {
+	ip := name.Name
+	log := log.WithLabels("controller", "ambient", "ip", ip)
+	log.Debugf("reconcile")
+
+	// Should be 0 or 1 pod, unless we are in a weird state
+	pod := PickBestPod(a.podIPIndex.Lookup(ip))
+
+	// Should be 0 or 1 entry, unless we are in a weird state
+	var entry *networking.WorkloadEntry
+	var configEntry *config.Config
+	if a.workloadEntriesIPIndex != nil {
+		entries := a.workloadEntriesIPIndex.Lookup(ip)
+		if len(entries) > 0 {
+			configEntry = entries[0]
+			entry = serviceentry.ConvertWorkloadEntry(*entries[0])
+		}
+	}
+
+	// We can have many slices (part of many services). Note each slice contains other IPs as well; we need to filter
+	epSlices := a.sliceIPIndex.Lookup(ip)
+	slices.SortFunc(epSlices, func(a, b *discoveryv1.EndpointSlice) bool {
+		// Newer slices win, could be an address moving from 1 to another
+		return a.CreationTimestamp.After(b.CreationTimestamp.Time)
+	})
+
+	if len(epSlices) == 0 && pod == nil && entry == nil {
+		log.Debugf("no endpoints or pod found")
+		return nil
+	}
+
+	// find the namespace for this workload
+	var namespace string
+	if pod != nil {
+		namespace = pod.Namespace
+	} else if entry != nil {
+		namespace = configEntry.Namespace
+	} else {
+		for _, slice := range epSlices {
+			namespace = slice.Namespace
+			break
+		}
+	}
+
+	vips := a.buildVIPsFromSlice(epSlices, namespace)
+	a.buildVIPsFromWorkloadEntry(entry, namespace, vips)
+
+	wl := &workloadapi.Workload{
+		Address:    parseIP(ip),
+		Namespace:  namespace,
+		Network:    a.controller.network.String(),
+		ClusterId:  a.controller.Cluster().String(),
+		VirtualIps: vips,
+	}
```

Comment:
nit; i think this is confusing to read and ensure all fields are set appropriately. id rather have this at the end of the function and then set other fields as necessary (with appropriate defaults) in a switch that comes before this block

Reply:
for example, we don't set a value for the Node in the WE case.. a field that ztunnel cares about during routing. perhaps we can set this to the workload entry address since there is a 1:1 mapping between node and WE address?